### PR TITLE
Fix include flags calculation for transitively pinned centrally managed dependencies

### LINKED ISSUE #12274: Include flags flow incorrectly to transitively pinned centrally managed dependencies
### NuGet Product Used

dotnet.exe

### Product Version

7.0.100

### Worked before?

no

### Impact

It bothers me. A fix would be nice

### Repro Steps & Context

The [graph flattening procedure](https://github.com/NuGet/NuGet.Client/blob/dfe13cc99d13de74f41e9cff35600a2c8958aac1/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/IncludeFlagUtils.cs#L67) doesn't work correctly for centrally managed transitive dependencies - in certain scenarios it includes too many assets.

https://github.com/marcin-krystianc/TestSolutions/tree/master/CpvmPrivateAssetsIssue4-IncludeFlags

App:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
    <CentralPackageTransitivePinningEnabled>true</CentralPackageTransitivePinningEnabled>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="..\Lib1\Lib1.csproj" />
  </ItemGroup>
</Project>
```
Lib1:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>	
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
    <CentralPackageTransitivePinningEnabled>true</CentralPackageTransitivePinningEnabled>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="..\Lib2\Lib2.csproj" IncludeAssets="Runtime" />
  </ItemGroup>
</Project>
```
Lib2:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
    <CentralPackageTransitivePinningEnabled>true</CentralPackageTransitivePinningEnabled>
  </PropertyGroup>  
  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" />
  </ItemGroup> 
</Project>
```
and `Directory.Packages.props`:
```
<Project>
    <ItemGroup>
        <PackageVersion Include="Newtonsoft.Json" Version="13.0.1" />
    </ItemGroup>
</Project>
```

The expected content of `centralTransitiveDependencyGroups` for App project is:
```
"centralTransitiveDependencyGroups": {
    ".NETStandard,Version=v2.0": {
      "Newtonsoft.Json": {
        "include": "Runtime",
        "version": "[13.0.1, )"
    }
  }
}
```

but .NET 7.0.100 produces:
```
"centralTransitiveDependencyGroups": {
  ".NETStandard,Version=v2.0": {
    "Newtonsoft.Json": {
      "include": "Runtime, Compile, Native, BuildTransitive",
      "suppressParent": "None",
      "version": "[13.0.1, )"
    }
  }
}
```

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by marcin-krystianc**
## Bug

Fixes: https://github.com/NuGet/Home/issues/12274

Regression? Last working version: no

## Description

When the include flags are calculated, only real (`LibraryDependencyReferenceType.Direct`) dependencies should be followed. The graph flattening algorithm should skip helper references (`LibraryDependencyReferenceType.None` and `LibraryDependencyReferenceType.Transitive`).

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by marcin-krystianc** [comment]
Next one is https://github.com/NuGet/NuGet.Client/pull/4952

**Review by jeffkl — DISMISSED** [review]

File: test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs [hunk 1]
```diff
@@ -1759,6 +1762,106 @@ public async Task IncludeType_DependenciesWithTargetsAsync()
...
+        {
+            // Arrange
+            using (var tmpPath = new SimpleTestPathContext())
+            {
+                var packageA = new SimpleTestPackageContext { Id = "PackageA", Version = "1.0.0", };
+                var logger = new TestLogger();
+                var project1Directory = new DirectoryInfo(Path.Combine(tmpPath.SolutionRoot, "Project1"));
+                var project2Directory = new DirectoryInfo(Path.Combine(tmpPath.SolutionRoot, "Project2"));
+                var project3Directory = new DirectoryInfo(Path.Combine(tmpPath.SolutionRoot, "Project3"));
+                var globalPackages = new DirectoryInfo(Path.Combine(tmpPath.WorkingDirectory, "globalPackages"));
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs]: SimpleTestPathContext already has a global packages folder and package source. 
You should be able to use those instead.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs [hunk 2]
```diff
@@ -1759,6 +1762,106 @@ public async Task IncludeType_DependenciesWithTargetsAsync()
...
+                var request = await ProjectTestHelpers.GetRequestAsync(restoreContext, project1Spec, project2Spec, project3Spec);
+
+                // Act
+                var command1 = new RestoreCommand(request);
+                var result1 = await command1.ExecuteAsync();
+                var lockFile1 = result1.LockFile;
+
+                // Assert
+                Assert.True(result1.Success);
+                Assert.Equal(includeFlags, lockFile1.CentralTransitiveDependencyGroups.Single().TransitiveDependencies.Single().IncludeType);
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs:1853]: This should affect the files selected as well correct?

Can we validate the targets section is correct as well?

**Review by nkolev92 — DISMISSED** [review]

File: test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs (hunk 2 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs:1853]: done 👍

File: test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs (hunk 1 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/IncludeTypeTests.cs]: done 👍

**Review by nkolev92 — APPROVED** [review]
Thanks for addressing the test feedback!

Great change.

**Review by jeffkl — APPROVED** [review]

**Comment by erdembayar** [comment]
I just requeued the failing Apex test job on CI pipeline.

**Review by erdembayar — APPROVED** [review]

**Pull request merged**